core-rust/node-common/src/scheduler.rs

```diff
@@ -62,190 +62,216 @@
  * permissions under this License.
  */
 
+use std::fmt::Display;
 use std::sync::Arc;
 
-use core::{
-    future::Future,
-    pin::Pin,
-    task::{Context, Poll},
-};
-use std::time::Duration;
+use prometheus::{HistogramVec, Registry};
+use std::time::{Duration, Instant};
 
 use tokio::runtime::Runtime;
 use tokio_util::sync::CancellationToken;
 
 use crate::locks::{LockFactory, Mutex};
-
-pub trait CancellableToken {
-    /// Be aware that cancellation is not an atomic operation. It is possible
-    /// for another thread running in parallel with a call to `cancel` to first
-    /// receive `true` from `is_cancelled` on one child node, and then receive
-    /// `false` from `is_cancelled` on another child node. However, once the
-    /// call to `cancel` returns, all child nodes have been fully cancelled.
-    fn cancel(&self);
+use crate::metrics::{new_timer_vec, opts, AtDefaultRegistryExt};
+
+/// A transient starter of periodically-executed tasks.
+///
+/// Important scheduler API note:
+/// The actual task-starting methods (e.g. [`Self::start_periodic()`]) *consume* the scheduler
+/// instance. This is by design, since we want to avoid 2 scheduled tasks of the same name (i.e.
+/// they could be confused in the log messages or the metrics). If you actually have a use-case for
+/// dynamically-created, unbounded, unidentified tasks, you can still achieve it by cloning
+/// the factory instance - just make sure to unconfigure any features that rely on unique names.
+#[derive(Clone)]
+pub struct Scheduler<'a> {
+    runtime: Option<&'a Runtime>,
+    name: String,
+    running_task_tracker: Option<&'a RunningTaskTracker>,
+    metrics: Option<SchedulerMetrics>,
 }
 
-/// Forward/delegate implementation of [`CancellableToken`] for Tokio's [`CancellationToken`]
-impl CancellableToken for CancellationToken {
-    fn cancel(&self) {
-        self.cancel();
+impl<'a> Scheduler<'a> {
+    /// Creates a new scheduler instance using the given tokio `runtime`.
+    /// The given `base_name` simply becomes the first segment of the constructed tasks' names' (see
+    /// [`Self::named()`]).
+    pub fn new(runtime: &'a Runtime, base_name: impl Display) -> Self {
+        Self {
+            runtime: Some(runtime),
+            name: base_name.to_string(),
+            running_task_tracker: None,
+            metrics: None,
+        }
     }
-}
 
-pub struct NoopCancellationFuture;
-
-impl Future for NoopCancellationFuture {
-    type Output = ();
-
-    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<()> {
-        Poll::Ready(())
+    /// Creates a mock instance for testing purposes.
+    /// The "scheduled" tasks will not actually be started.
+    pub fn for_testing() -> Self {
```

Comment:
_(note from author)_

Obviously I liked the explicit `NoopScheduler` 10x more (than `runtime: None` 🤦 ).
But I was not able to make `trait Scheduler` work together with the "immutable builder pattern" - basically I had the same problem with `LockFactory`: the returned types became too wild and nested (or required `Box<dyn ...>` everywhere).